compute_endpoint/globus_compute_endpoint/engines/high_throughput/engine.py

```diff
@@ -0,0 +1,924 @@
+            self._queue_management_thread = threading.Thread(
+                target=self._queue_management_worker, name="Queue-Management"
+            )
+            self._queue_management_thread.daemon = True
+            self._queue_management_thread.start()
+            log.debug("Started queue management thread")
+
+        else:
+            log.debug("Management thread already exists, returning")
+
+    def hold_worker(self, worker_id):
+        """Puts a worker on hold, preventing scheduling of additional tasks to it.
+
+        This is called "hold" mostly because this only stops scheduling of tasks,
+        and does not actually kill the worker.
+
+        Parameters
+        ----------
+
+        worker_id : str
+            Worker id to be put on hold
+        """
+        c = self.command_client.run(f"HOLD_WORKER;{worker_id}")
+        log.debug(f"Sent hold request to worker: {worker_id}")
+        return c
+
+    def send_heartbeat(self):
+        log.warning("Sending heartbeat to interchange")
+        msg = Heartbeat(endpoint_id="")
+        self.outgoing_q.put(msg.pack())
+
+    def wait_for_endpoint(self):
+        heartbeat = self.command_client.run(HeartbeatReq())
+        log.debug("Attempting heartbeat to interchange")
+        return heartbeat
+
+    @property
+    def outstanding(self):
+        outstanding_c = self.command_client.run("OUTSTANDING_C")
+        log.debug(f"Got outstanding count: {outstanding_c}")
+        return outstanding_c
+
+    @property
+    def connected_workers(self):
+        workers = self.command_client.run("MANAGERS")
+        log.debug(f"Got managers: {workers}")
+        return workers
+
+    def _submit(
+        self,
+        func: t.Callable,
+        *args: t.Any,
+        task_id: uuid.UUID = None,
+        **kwargs: t.Any,
+    ) -> Future:
+        self._task_counter += 1
+        if task_id is None:
+            s_task_id = str(self._task_counter)
+        else:
+            s_task_id = str(task_id)
+
+        ser_payload = fx_serializer.serialize((func, args, kwargs))
+        payload = Task(s_task_id, "RAW", ser_payload)
+        self.tasks[s_task_id] = HTEXFuture(self, task_id=s_task_id)
+        self.tasks[s_task_id].task_id = s_task_id
+        assert self.outgoing_q
+        self.outgoing_q.put(payload.pack())
+
+        return self.tasks[s_task_id]
+
+    def submit(self, task_id: uuid.UUID, packed_task: bytes) -> Future:
+        """Submits a messagepacked.Task for execution
+
+        Parameters
+        ----------
+        Packed Task (messages.Task) - A packed Task object which contains task_id,
+        container_id, and serialized fn, args, kwargs packages.
+
+        Returns:
+              Submit status
+        """
+        log.debug(f"Submitting task:{task_id}")
+
+        if self._executor_bad_state.is_set():
+            # If the flag is set the exception body must exist
+            raise self._executor_exception  # type: ignore
+
```

Comment:
`_executor_exception` is a write-once variable, yeah?  If so, seems like there's no need for `self._executor_bad_state`.  Is it a vestigial thought-process from prior to some refactor?  The event doesn't appear to be shared anywhere else.  Thus, this could be simplified (and thus, not type-ignored) to:
```python
if self._executor_exception:
    # Truthy, so it's *not* None, and thus mypy won't complain:
    raise self._executor_exception
```

Reply:
You are probably right. I'd like to keep this for later since I'm not completely sure about this.